flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -417,11 +417,41 @@ public void cancel() {
 	@Override
 	public final boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
 		final Optional<ExecutionVertexID> executionVertexId = getExecutionVertexId(taskExecutionState.getID());
-		if (executionVertexId.isPresent()) {
-			executionGraph.updateState(taskExecutionState);
-			updateTaskExecutionStateInternal(executionVertexId.get(), taskExecutionState);
+
+		boolean updateSuccess = executionGraph.updateState(taskExecutionState);
+
+		if (updateSuccess) {
+			if (isNotifiable(executionVertexId.get(), taskExecutionState)) {
```

Comment:
I think now it's better to add `checkState(executionVertexId.isPresent());` prior to this line.

Reply:
done.

Reply:
Just noticed that we are addressing the last comment.